### PR TITLE
UE: More Bits!

### PR DESCRIPTION
**Pull request by Hendiadyoin1**
* Adds support for 128/256 bit reads and writes
* ~Adds support for Packed Data and operations upon them~
* Adds FLD_RM80 and FSTP_RM80 instruction support

Known Issues:
* ~u256 is not constexpr'nable, at least not in a neat way due to u128 not being "trivially constructible"~ u128 and u256 are not trivially contructable
* narrowing conversions from u128/u256 down are not marked by the compiler
* ~the name of the Packed data containers is not that good (is it even in the right place?)~ deleted for now
* We could do two reads/writes for f80 instead of one too big one

Side Note:
* long double's are weird with their supposed size

### CODE REVIEW

**Review by bcoles — COMMENTED** [review]
Copyright headers on all new files say 2020 (not 2021).

**Comment by Hendiadyoin1** [comment]
We now support printing the containers and a whole lot more actions with them

**Comment by Hendiadyoin1** [comment]
regarding the 4th known issue (to big reads), One Idea would be:
``` cpp
// writing
{       
    [...]
    auto address = insn.modrm().resolve(this,insn);
    u8 data[sizeof(long double)];
    *(long double*)data = fpu_pop();
    
    write_memory64(address, {data, 0x01010101'01010101ULL});
    if constexpr (sizeof(long double) == 12)
        write_memory32(address, {*(u32*)&data[8], 0x01010101});
    else if constexpr (sizeof(long double) == 16)
        write_memory64(address, {*(u64*)&data[8],0x01010101'01010101ULL});
    else VERIFY_NOT_REACHED();
}
//reading
{
    auto address = insn.modrm().resolve(this,insn);

    u8 data[sizeof(long double)];
    *(u64*)data =  read_memory64(address).value();
    if constexpr (sizeof(long double) == 12)
        *(u32*)&data[8] =  read_memory32(address).value();
    else if constexpr (sizeof(long double) == 16)
        *(u64*)&data[8] =  read_memory64(address).value();
    else VERIFY_NOT_REACHED();

    fpu_push(*(long double*)data);
}
```

File: Userland/DevTools/UserspaceEmulator/SoftCPU.cpp [hunk 1]
```diff
@@ -1834,7 +1834,18 @@ void SoftCPU::FNDISI(const X86::Instruction&) { TODO_INSN(); }
 void SoftCPU::FNCLEX(const X86::Instruction&) { TODO_INSN(); }
 void SoftCPU::FNINIT(const X86::Instruction&) { TODO_INSN(); }
 void SoftCPU::FNSETPM(const X86::Instruction&) { TODO_INSN(); }
-void SoftCPU::FLD_RM80(const X86::Instruction&) { TODO_INSN(); }
+
+void SoftCPU::FLD_RM80(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+
+    // long doubles can be up to 128 bits wide in memmory for reasons (alignment) and only uses 80 bits of precision
```
Comment (awesomekling) [inline Userland/DevTools/UserspaceEmulator/SoftCPU.cpp:1842]: memmory -> memory

File: Userland/DevTools/UserspaceEmulator/SoftCPU.cpp [hunk 2]
```diff
@@ -1834,7 +1834,18 @@ void SoftCPU::FNDISI(const X86::Instruction&) { TODO_INSN(); }
...
+    // bit_cast seems to see the 12 bytes, but godbolt says 16 bytes in x86-64 or 12 with -m32
```
Comment (awesomekling) [inline Userland/DevTools/UserspaceEmulator/SoftCPU.cpp:1843]: godbolt is a website. Are you saying that a compiler is telling you something? Please be more clear.

File: Userland/DevTools/UserspaceEmulator/SoftCPU.cpp [hunk 3]
```diff
@@ -1834,7 +1834,18 @@ void SoftCPU::FNDISI(const X86::Instruction&) { TODO_INSN(); }
...
+    // and writing a 12 byte read just for this would be ineficient
```
Comment (awesomekling) [inline Userland/DevTools/UserspaceEmulator/SoftCPU.cpp:1844]: ineficient -> inefficient
Also, why would it be inefficient?

File: Userland/DevTools/UserspaceEmulator/SoftCPU.cpp [hunk 4]
```diff
@@ -1871,7 +1882,22 @@ void SoftCPU::FCOMI(const X86::Instruction& insn)
...
-void SoftCPU::FSTP_RM80(const X86::Instruction&) { TODO_INSN(); }
+void SoftCPU::FSTP_RM80(const X86::Instruction& insn)
+{
+    if (insn.modrm().is_register()) {
+        fpu_set(insn.modrm().register_index(), fpu_pop());
+    } else {
+        // FIXME: Respect shadow values
+        // long doubles can be up to 128 bits wide in memmory for reasons (alignment) and only uses 80 bits of precision
+        // bit_cast seems to see the 12 bytes, but godbolt says 16 bytes in x86-64 or 12 with -m32
+        // and writing a 12 byte write just for this would be ineficient
```
Comment (awesomekling) [inline Userland/DevTools/UserspaceEmulator/SoftCPU.cpp:1893]: Same comments as above.

File: Userland/DevTools/UserspaceEmulator/SoftCPU.cpp [hunk 5]
```diff
@@ -1871,7 +1882,22 @@ void SoftCPU::FCOMI(const X86::Instruction& insn)
...
+        // and for that reason we need to read first to preserve the other few bytes,
+        // although it taints the memmory graph
```
Comment (awesomekling) [inline Userland/DevTools/UserspaceEmulator/SoftCPU.cpp:1895]: memmory -> memory

File: AK/BitCast.h [hunk 6]
```diff
@@ -34,7 +40,7 @@ inline T bit_cast(const U& a)
     static_assert(sizeof(T) == sizeof(U));
 
     T result;
-    __builtin_memcpy(&result, &a, sizeof(T));
+    memcpy((void*)&result, &a, sizeof(T));
```
Comment (awesomekling) [inline AK/BitCast.h:43]: Why is this necessary? Does it change the generated code at all?

**Review by awesomekling — CHANGES_REQUESTED** [review]
What is IdentMap? What is "Ident" in the X86 context?
Please explain what you are doing :)

File: Userland/Libraries/LibX86/Types/Formatter.cpp [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    if (value.high() == 0) {
+        AK::Formatter<u64> formatter { *this };
+        return formatter.format(builder, value.low());
+    }
+
+    if (m_precision.has_value())
+        VERIFY_NOT_REACHED();
+
+    if (m_mode == Mode::Pointer) {
+        // this is way to beig for a pointer
```
Comment (bcoles) [inline Userland/Libraries/LibX86/Types/Formatter.cpp:47]: big?

```suggestion
        // this is way to big for a pointer
```

File: Userland/Libraries/LibX86/Types/Formatter.cpp [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    if (m_mode == Mode::Binary) {
+        base = 2;
+    } else if (m_mode == Mode::BinaryUppercase) {
+        base = 2;
+        upper_case = true;
+    } else if (m_mode == Mode::Octal) {
+        base = 8;
+    } else if (m_mode == Mode::Decimal || m_mode == Mode::Default) {
+        // FIXME: implement this
+        VERIFY_NOT_REACHED();
```
Comment (bcoles) [inline Userland/Libraries/LibX86/Types/Formatter.cpp:62]:
```suggestion
        TODO();
```

File: Userland/Libraries/LibX86/Types/Formatter.cpp [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    if (value.high() == 0) {
+        AK::Formatter<u128> formatter { *this };
+        return formatter.format(builder, value.low());
+    }
+
+    if (m_precision.has_value())
+        VERIFY_NOT_REACHED();
+
+    if (m_mode == Mode::Pointer) {
+        // this is way to beig for a pointer
```
Comment (bcoles) [inline Userland/Libraries/LibX86/Types/Formatter.cpp:92]:
```suggestion
        // this is way to big for a pointer
```

File: Userland/Libraries/LibX86/Types/Formatter.cpp [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    if (m_mode == Mode::Binary) {
+        base = 2;
+    } else if (m_mode == Mode::BinaryUppercase) {
+        base = 2;
+        upper_case = true;
+    } else if (m_mode == Mode::Octal) {
+        base = 8;
+    } else if (m_mode == Mode::Decimal || m_mode == Mode::Default) {
+        // FIXME: implement this
+        VERIFY_NOT_REACHED();
```
Comment (bcoles) [inline Userland/Libraries/LibX86/Types/Formatter.cpp:107]:
```suggestion
        TODO();
```

File: Userland/Libraries/LibX86/Types/IdentView.h [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    template<Integral R, Integral U>                                       \
+    ALWAYS_INLINE IdentMap<R> operator op(const IdentView<U>& other) const \
+    {                                                                      \
+        VERIFY(m_size == other.size());                                    \
+        IdentMap<R> ret(m_size);                                           \
+        for (off_t i = 0; i < m_size; ++i) {                               \
+            ret[i] = m_values[i] op other[i];                              \
+        }                                                                  \
+        return ret;                                                        \
+    }                                                                      \
```
Comment (bcoles) [inline Userland/Libraries/LibX86/Types/IdentView.h:64]: Both `IdentView.h` and `IdentMap.h` are very similar and `#include` each other.

These highlighted functions are identical in both files. Are these supposed to be `ALWAYS_INLINE IdentView<R> ...` rather than `ALWAYS_INLINE IdentMap<R> ...` ?

File: Userland/Libraries/LibX86/Types/u128.h [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+    ALWAYS_INLINE u128 operator|(const u128& other) const
+    {
+        return { low() | other.low(), high() | other.high() };
+    }
+    ALWAYS_INLINE u128 operator^(const u128& other) const
+    {
+        return { low() ^ other.low(), high() ^ other.high() };
+    }
+
+    // bitwise asign
```
Comment (bcoles) [inline Userland/Libraries/LibX86/Types/u128.h:220]:
```suggestion
    // bitwise assign
```

File: Userland/Libraries/LibX86/Types/u256.h [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+    ALWAYS_INLINE u256 operator|(const u256& other) const
+    {
+        return { low() | other.low(), high() | other.high() };
+    }
+    ALWAYS_INLINE u256 operator^(const u256& other) const
+    {
+        return { low() ^ other.low(), high() ^ other.high() };
+    }
+
+    // bitwise asign
```
Comment (bcoles) [inline Userland/Libraries/LibX86/Types/u256.h:228]:
```suggestion
    // bitwise assign
```

File: Userland/Libraries/LibX86/Types/IdentView.h (hunk 11 above)
Comment (Hendiadyoin1) [reply Userland/Libraries/LibX86/Types/IdentView.h:64]: In this  its a map, because it uses its own copy of data, instead of acting upon data provided to it

File: AK/BitCast.h (hunk 6 above)
Comment (Hendiadyoin1) [reply AK/BitCast.h:43]: haven't checked that, but I just happened  to stumble on that while trying to use bit_cast's for long double conversion and I thought the "zen" was to try write every method on our own, and not rely on compilers and their libraries if possible

the (void*) cast is useless though

File: Userland/DevTools/UserspaceEmulator/SoftCPU.cpp (hunk 3 above)
Comment (Hendiadyoin1) [reply Userland/DevTools/UserspaceEmulator/SoftCPU.cpp]: code wise  
it would be another function that would have to be implemented across multiple classes like the other reads/writes  
in a previous comment I have described a way to do it in 2 writes dependent on long double size
In the newer version its described and handled better

File: Userland/DevTools/UserspaceEmulator/SoftCPU.cpp (hunk 2 above)
Comment (Hendiadyoin1) [reply Userland/DevTools/UserspaceEmulator/SoftCPU.cpp]: Changed it to refer to gcc's behavior

**Comment by Hendiadyoin1** [comment]
> 
> 
> What is IdentMap? What is "Ident" in the X86 context?
> Please explain what you are doing :)


So these are supposed to be views on packed data, with which you can perform the same arythmatic across all values at once.
I thought it may be a good way to implement the structure backing ymm and xmm vector registers

File: Userland/Libraries/LibX86/Types/u256.h (hunk 13 above)
Comment (Hendiadyoin1) [reply Userland/Libraries/LibX86/Types/u256.h]: should be fixed

File: Userland/Libraries/LibX86/Types/u128.h (hunk 12 above)
Comment (Hendiadyoin1) [reply Userland/Libraries/LibX86/Types/u128.h]: should be fixed

File: Userland/Libraries/LibX86/Types/Formatter.cpp (hunk 10 above)
Comment (Hendiadyoin1) [reply Userland/Libraries/LibX86/Types/Formatter.cpp]: should be fixed

File: Userland/Libraries/LibX86/Types/Formatter.cpp (hunk 9 above)
Comment (Hendiadyoin1) [reply Userland/Libraries/LibX86/Types/Formatter.cpp]: should be fixed

File: Userland/Libraries/LibX86/Types/Formatter.cpp (hunk 8 above)
Comment (Hendiadyoin1) [reply Userland/Libraries/LibX86/Types/Formatter.cpp]: should be fixed

File: Userland/Libraries/LibX86/Types/Formatter.cpp (hunk 7 above)
Comment (Hendiadyoin1) [reply Userland/Libraries/LibX86/Types/Formatter.cpp]: should be fixed

File: Userland/DevTools/UserspaceEmulator/SoftCPU.cpp (hunk 5 above)
Comment (Hendiadyoin1) [reply Userland/DevTools/UserspaceEmulator/SoftCPU.cpp]: should be fixed

File: Userland/DevTools/UserspaceEmulator/SoftCPU.cpp (hunk 4 above)
Comment (Hendiadyoin1) [reply Userland/DevTools/UserspaceEmulator/SoftCPU.cpp]: same answer as above.

File: Userland/DevTools/UserspaceEmulator/SoftCPU.cpp (hunk 1 above)
Comment (Hendiadyoin1) [reply Userland/DevTools/UserspaceEmulator/SoftCPU.cpp]: should be fixed

File: AK/BitCast.h (hunk 6 above)
Comment (Hendiadyoin1) [reply AK/BitCast.h:43]: And looking at [this](https://godbolt.org/z/9jqd5GfWc) it seems that buildins are better due to being optimized better to just to movs instead of a rep movsb and allways inlined, maybe we can apply some of these optimizations too in out libC implementation

File: Userland/Libraries/LibX86/Types/IdentView.h (hunk 11 above)
Comment (Hendiadyoin1) [reply Userland/Libraries/LibX86/Types/IdentView.h:64]: And when doing operations we do not want to alter the data directly so its a Map there

File: AK/BitCast.h (hunk 6 above)
Comment (Hendiadyoin1) [reply AK/BitCast.h:43]: guess I'll drop that change for now

File: Userland/DevTools/UserspaceEmulator/ValueWithShadow.h [hunk 14]
```diff
@@ -24,11 +24,16 @@
...
 #include <AK/Format.h>
 #include <AK/Platform.h>
+#include <LibX86/Types.h>
+#include <string.h>
 
-#pragma once
-
+#define INITIALIZED_64 0x01010101'01010101LLU
+#define INITIALIZED_128 u128(INITIALIZED_64, INITIALIZED_64)
+#define INITIALIZED_256 u256(INITIALIZED_128, INITIALIZED_128)
```
Comment (awesomekling) [inline Userland/DevTools/UserspaceEmulator/ValueWithShadow.h:36]: Can we make these `constexpr` values instead?

File: Userland/DevTools/UserspaceEmulator/SoftCPU.cpp [hunk 15]
```diff
@@ -1871,7 +1882,23 @@ void SoftCPU::FCOMI(const X86::Instruction& insn)
...
+void SoftCPU::FSTP_RM80(const X86::Instruction& insn)
+{
+    if (insn.modrm().is_register()) {
+        fpu_set(insn.modrm().register_index(), fpu_pop());
+    } else {
+        // FIXME: Respect shadow values
+        // long doubles can be up to 128 bits wide in memory for reasons (alignment) and only uses 80 bits of precision
+        // gcc uses 12 byte in 32 bit and 16 byte in 64 bit mode
+        // so in the 32 bit case we have to read first, to not override data on the overly big write
+        u128 f80();
```
Comment (awesomekling) [inline Userland/DevTools/UserspaceEmulator/SoftCPU.cpp:1894]: This is a very strange declaration. please use brace initialization instead, i.e `u128 f80 {};`

**Review by awesomekling — CHANGES_REQUESTED** [review]
> > What is IdentMap? What is "Ident" in the X86 context?
> > Please explain what you are doing :)
> 
> So these are supposed to be views on packed data, with which you can perform the same arythmatic across all values at once.
> I thought it may be a good way to implement the structure backing ymm and xmm vector registers
> 
> I know the name is bad how about "DataVector[...]"

Since you are not using them for anything here, it's just dead code right now, so I think it's better to delay introducing them until they're actually needed for something.

File: Userland/DevTools/UserspaceEmulator/SoftCPU.cpp (hunk 15 above)
Comment (Hendiadyoin1) [reply Userland/DevTools/UserspaceEmulator/SoftCPU.cpp:1894]: I thought i changed that, but that may explain the build error

File: Userland/DevTools/UserspaceEmulator/ValueWithShadow.h (hunk 14 above)
Comment (Hendiadyoin1) [reply Userland/DevTools/UserspaceEmulator/ValueWithShadow.h:36]: the byte access in in u256 would be a cast of `this` to the corresponding type, but yes

this will be fixed in the next commit iteration

**Comment by Hendiadyoin1** [comment]
> 
> 
> > > What is IdentMap? What is "Ident" in the X86 context?
> > > Please explain what you are doing :)
> > 
> > 
> > So these are supposed to be views on packed data, with which you can perform the same arythmatic across all values at once.
> > I thought it may be a good way to implement the structure backing ymm and xmm vector registers
> > I know the name is bad how about "DataVector[...]"
> 
> Since you are not using them for anything here, it's just dead code right now, so I think it's better to delay introducing them until they're actually needed for something.

Then I'll stash them away for now

**Comment by Hendiadyoin1** [comment]
Was that everything?
